sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala

```diff
@@ -68,8 +70,11 @@ class InMemoryFileIndex(
   refresh0()
 
   override def partitionSpec(): PartitionSpec = {
+    if (userSpecifiedPartitionSpec.isDefined) {
+      cachedPartitionSpec = userSpecifiedPartitionSpec.get
+    }
     if (cachedPartitionSpec == null) {
-      cachedPartitionSpec = inferPartitioning()
+        cachedPartitionSpec = inferPartitioning()
```

Comment:
why not put the `if (userSpecifiedPartitionSpec.isDefined)` inside this if?

Reply:
The test failed when I put it in this if : 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116149/testReport
But it may be an irrelevant error. Will retry it.